torch/jit/annotations.py

```diff
@@ -292,11 +292,14 @@ def try_ann_to_type(ann, loc):
         value = try_ann_to_type(ann.__args__[1], loc)
         return DictType(key, value)
     if is_optional(ann):
+        print(ann)
+        print(ann.__args__)
         if issubclass(ann.__args__[1], type(None)):
-            valid_type = try_ann_to_type(ann.__args__[0], loc)
+            contained = ann.__args__[0]
         else:
-            valid_type = try_ann_to_type(ann.__args__[1], loc)
-        assert valid_type, "Unsupported annotation {} could not be resolved.".format(repr(ann))
+            contained = ann.__args__[1]
+        valid_type = try_ann_to_type(contained, loc)
+        assert valid_type, "Unsupported annotation {} could not be resolved.".format(repr(contained))
```

Comment:
I think this is a good change to make message clearer.  Just one minor suggestion, what do you think about making this message a bit more verbose? Something like 
```
unsupported annotation Union[T, None] could not be resolved, because T could not be resolved.
```

Reply:
Yeah, good point, Optional[T] is better than Union[T, None]. I just wanted to make sure users don't get confused the other way : "why is JIT complaining about type being T when I clearly said it is Optional[T]"

Reply:
```
>>> from typing import Optional
>>> str(Optional[int])
'typing.Union[int, NoneType]'
```
I guess we can leave it as `Union` but add the extra phrasing you suggested.